server/src/main/java/org/elasticsearch/index/shard/IndexingOperationListener.java

```diff
@@ -62,6 +64,15 @@ default void postDelete(ShardId shardId, Engine.Delete delete, Engine.DeleteResu
      */
     default void postDelete(ShardId shardId, Engine.Delete delete, Exception ex) {}
 
+    /**
+     * Called when a {@link org.elasticsearch.action.bulk.TransportShardBulkAction} is about to perform index and/or delete operation(s)
+     * on a primary shard.
+     *
+     * @param indexShard the shard the bulk is about to be performed on
+     * @param proceedListenerSupplier call this immediately to get a listener which must be completed so that the bulk can proceed.
+     */
+    default void preBulkOnPrimary(IndexShard indexShard, Supplier<ActionListener<Void>> proceedListenerSupplier) {}
```

Comment:
We should add a comment here that this gets called from a transport thread and therefore the listener should be really lightweight and don't block the thread.